packages/records/lib/models/records.ts

```diff
@@ -216,53 +216,97 @@ export async function upsert({
+                // Retry if deadlock detected
                 // https://www.postgresql.org/docs/current/mvcc-serialization-failure-handling.html
-                const upserting = () =>
-                    trx.from<FormattedRecord>(RECORDS_TABLE).insert(encryptedRecords).onConflict(['connection_id', 'external_id', 'model']).merge();
-                await retry(upserting, {
-                    maxAttempts: 3,
-                    delayMs: 500,
-                    retryIf: (res) => {
-                        if ('code' in res) {
-                            const errorCode = (res as { code: string }).code;
-                            return errorCode === '40P01'; // deadlock_detected
+                const withRetry = async <T>(query: Knex.QueryBuilder<any, T>) => {
+                    return retry(() => query, {
+                        maxAttempts: 3,
+                        delayMs: 500,
+                        retryIf: (res) => {
+                            if ('code' in res) {
+                                const errorCode = (res as { code: string }).code;
+                                return errorCode === '40P01'; // deadlock_detected
+                            }
+                            return false;
                         }
-                        return false;
-                    }
-                });
+                    });
+                };
 
-                const delta = chunkSummary.addedKeys.length - (chunkSummary.deletedKeys?.length ?? 0);
-                if (delta !== 0) {
-                    await trx
-                        .from(RECORD_COUNTS_TABLE)
-                        .insert({
-                            connection_id: connectionId,
-                            model,
-                            environment_id: environmentId,
-                            count: delta
-                        })
-                        .onConflict(['connection_id', 'environment_id', 'model'])
-                        .merge({
-                            count: trx.raw(`${RECORD_COUNTS_TABLE}.count + EXCLUDED.count`),
-                            updated_at: trx.fn.now()
-                        });
+                // we need to know which records were updated, deleted, undeleted or unchanged
+                // we achieve this by comparing the records data_hash and deleted_at fields before and after the update
+                const existingIds = chunk.map((r) => r.external_id);
+                const updateQuery = trx
+                    .with('existing', (qb) => {
+                        qb.select('external_id', 'data_hash', 'deleted_at')
+                            .from(RECORDS_TABLE)
+                            .where({
+                                connection_id: connectionId,
+                                model
+                            })
+                            .whereIn('external_id', existingIds)
+                            .forUpdate();
```

Comment:
note: this is similar to an advisory lock in the end since you lock all the rows. 
I'm wondering about:
- the perf of locking all rows like this vs one lock.
- if it's working since you can only lock rows that exist. if a webhook is running concurrently the two operations could potentially execute at the same time, would only be an issue with record creation maybe I don't know

Reply:
It is locking indeed. There is no real way around it but as you mention an explicit advisory lock could be subject to deadlock or connection pool congestion. 
Locking the rows that are about to be modified is not a bad idea I think

Regarding you second point, yes you are right. In case of concurrent records creation, the first one to occur is going to win but whatever strategy we decide to implement we always gonna have to make a choice between one or the other

Reply:
I think my point boils down to this: it will use the same number of connections since it's also a global lock but in a less explicit way

Reply:
I am fine using an advisory lock if this is what you prefer